### PR TITLE
Added basic attributes support

### PR DESCRIPTION
**Pull request by DavertMik**
Added attributes for Cest format (some of them also work for Test)

Usage example (updated):

```php
    #[Incomplete]
    #[Skip]
    #[Group('ue1')]
    #[Example([1, 1], [2, 2])]
    #[Depends('test1')]
    #[Before('test1')]
    #[Before('test2')]
    public function myTest($I) 
    {
    }
```

### CODE REVIEW

**Review by TavoNiievez — COMMENTED** [review]
I'm not sure modifying `src/Codeception/Util/Annotation.php` is the best way to add this.
Is it possible to create a separate class just for handling the attributes?
Are you sure you want attributes to start with lowercase?

It looks good, what doesn't convince me is where it was added.

**Comment by Naktibalda** [comment]
> While not strictly required it is recommended to create an actual class for every attribute.

https://www.php.net/manual/en/language.attributes.classes.php

PHPUnit has 49 attribute classes: https://github.com/sebastianbergmann/phpunit/tree/main/src/Framework/Attributes

**Comment by DavertMik** [comment]
@TavoNiievez 
> I'm not sure modifying src/Codeception/Util/Annotation.php is the best way to add this.
Is it possible to create a separate class just for handling the attributes?

I think it is the right way as 90% of logic is the same, to get the reflection for a method and parse out data from it.
At this point, attributes are just syntax sugar for annotations. I don't see why we need a new class for it. Maybe rename the current one?

> Are you sure you want attributes to start with lowercase?

Good point, I'm also not so happy to set a root namespace for all of them... 
Lowercase was chosen to follow annotations but I agree that this can be updated.

@Naktibalda 

> While not strictly required it is recommended to create an actual class for every attribute.

Well, we can add it. For instance, this may differentiate Codeception's attributes from PHPUnit's attributes, as some of them are quite different like `Before`. I don't like one extra Import line here but the overall idea looks good. We don't have 49 attributes so it should not be a problem to add them.

File: src/Codeception/Util/Annotation.php [hunk 1]
```diff
@@ -119,13 +124,39 @@ public function fetch(string $annotation): ?string
...
+            return $attr->getArguments();
+        }
         $docBlock = (string)$this->currentReflectedItem->getDocComment();
         if (preg_match_all(sprintf(self::$regex, $annotation), $docBlock, $matched)) {
             return $matched[1];
         }
         return [];
     }
 
+    public function attributes(): array
```
Comment (TavoNiievez) [inline src/Codeception/Util/Annotation.php:138]: Awesome!

**Review by TavoNiievez — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

File: src/Codeception/Util/Annotation.php [hunk 2]
```diff
@@ -119,13 +124,39 @@ public function fetch(string $annotation): ?string
...
+    {
+        $attrs = $this->currentReflectedItem->getAttributes();
+        $attrs = array_filter($attrs);
+        $attrs = array_filter($attrs, fn (\ReflectionAttribute $a) => str_starts_with($a->getName(), 'Codeception\\Attribute\\'));
+        return $attrs;
+    }
+
+    public function attribute($name): ?\ReflectionAttribute
```
Comment (TavoNiievez) [inline src/Codeception/Util/Annotation.php:146]: `string $name`

File: src/Codeception/Test/Metadata.php [hunk 3]
```diff
@@ -196,6 +196,20 @@ public function setParamsFromAnnotations($annotations): void
         }
     }
 
+    public function setParamsFromAttributes($attributes): void
```
Comment (TavoNiievez) [inline src/Codeception/Test/Metadata.php:199]: `array $attributes`

**Comment by mbrodala** [comment]
Thanks a lot for this, especially the now unified `Examples` attribute is pure gold. :star_struck: 

If only there was a suitable Rector ... :sweat_smile:

**Pull request merged**